File: plclite.py
#!/usr/bin/python

"""PLC - lite for running logic on Raspberry pi"""
VERSION = '0.0'

USAGE = '''Usage: plclite [-p config file]
    Options:
    -h displays this help message
    -p uses a program fle";

'''

PROGRAM = 'program.ld'
STEP = 1000
INPUTS = [16,20,21]
OUTPUTS = [3]

import sys
sys.path.append('build-lib')

# librelogic.py
from _librelogic import ffi, lib
hardware_conf = ffi.new("struct config_gpiod *",
{   
    'chipname' : ffi.new("char[]", "gpiochip0".encode('ascii')),
    'in_lines' : ffi.new("uint32_t[]", INPUTS),
    'in_size'  : len(INPUTS),
    'out_lines': ffi.new("uint32_t[]", OUTPUTS),
    'out_size' : len(OUTPUTS),
    'label' : ffi.new("char[]", "gpiod".encode('ascii')),
}
)

def dump(plc):
    instr = "In:  "  
    for i in range(0,len(INPUTS)):
        instr = instr + str(lib.plc_get_di_val(plc, i)) + " "
    print(instr);

    qstr = "Out: "  
    for q in range(0,len(OUTPUTS)):
        qstr = qstr + str(lib.plc_get_dq_val(plc, q)) + " "
    print(qstr);
        

def logic(prog):
    hw_gpiod = lib.plc_get_hardware(lib.HW_GPIOD); 
    hw_gpiod.configure(hardware_conf);
    plc = lib.plc_new( 1, 1, 0, 0, 1, 0, 2, 0, STEP, hw_gpiod); 
    
    return lib.plc_load_program_file(ffi.new("char[]", prog.encode('ascii')), plc);

def main(args=None):
    import getopt

    p_file = PROGRAM
    
    if args is None:
        args = sys.argv[1:]
    try:
        opts, args = getopt.gnu_getopt(args, 'hvp:',
                                       ['help', 'version',
                                        'program='
                                       ])
        for o,a in opts:
            if o in ('-h', '--help'):
                print(USAGE)
                return 0
            elif o in ('-v', '--version'):
                print(VERSION)
                return 0
            elif o in ('-p', '--program'):
                p_file = a
            
    except getopt.GetoptError:
        e = sys.exc_info()[1]     # current exception
        sys.stderr.write(str(e)+"\n")
        sys.stderr.write(USAGE+"\n")
        return 1
        
    plc = lib.plc_start(logic(p_file))
        
    while(lib.plc_is_running(plc) == 1):
        plc = lib.plc_func(plc)
        if lib.plc_is_updated(plc) > 0:
            dump(lib.plc_reset_update(plc))

if __name__ == '__main__':
    sys.exit(main())

File: src/plclite.c
#include <stdio.h>
#include <unistd.h>
#include <ctype.h>
#include <signal.h>

#include <plclib.h>
      
#include "plclite.h" 
#ifdef SIM
//SIM configuration
#endif //SIM

#ifdef COMEDI
//COMEDI configuration

#endif //COMEDI

#ifdef USPACE
//uspace configuration
#endif //USPACE

#ifdef GPIOD
int Inputs[N_GPIO_IN] = {16,20,21};
int Outputs[N_GPIO_OUT] = {3};
struct config_gpiod GpiodConf = {
    .chipname = "gpiochip0",
    .in_lines = Inputs,
    .in_size = N_GPIO_IN,
    .out_lines = Outputs,
    .out_size = N_GPIO_OUT,
};
#endif //GPIOD

const char * Usage = "Usage: plclite [-p config file] \n \
    Options:\n \
    -h displays this help message\n \
    -p uses a program fle";
plc_t Plc;

void dump(){
    printf("\n%s %d\n" , "PLC status:", Plc->status);

    int i = 0;
    printf("\n%s\n" , "inputs:");
    while(i <  N_GPIO_IN){
        printf("%d", Plc->di[i++].I);
    }

    i = 0;
    printf("\n%s\n" , "outputs:");
    while(i <  N_GPIO_OUT){
        printf("%d", Plc->dq[i++].Q);
    }

    i = 0;
    printf("\n%s\n" , "memory regs:");
    while(i <  N_MEM){
        printf("%ld ", Plc->m[i++].V);
    }

    i = 0;
    printf("\n%s\n" , "timers:");
    while(i <  N_TIM){
        printf("%ld %d \n", Plc->t[i].V, Plc->t[i++].Q);
    }
}

/**
* @brief graceful shutdown
*/
void sigkill() {
    printf("%s\n",  "program interrupted");
    plc_clear(Plc);
    exit(0);
}


int main(int argc, char **argv)
{
    int prog = 0;
    char * progstr = PROGRAM;
    char * cvalue = NULL;
    opterr = 0;
    int c;

    signal(SIGINT, sigkill);
    signal(SIGTERM, sigkill);

    while ((c = getopt (argc, argv, "hp:")) != -1){
        switch (c) {
        case 'h':
         printf("%s\n", Usage);
         return 1;
        break;
        case 'p':
        cvalue = optarg;
        break;
        case '?':
         printf("%s\n", Usage);
        if (optopt == 'p'){
             printf( 
            "Option -%c requires an argument\n", optopt);
        } else if (isprint (optopt)){
            printf( 
            "Unknown option `-%c'\n", optopt);
        } else{
            printf( 
            "Unknown option character `\\x%x'\n",
            optopt);
        }
        return 1;
        default:
           return -1;
        }
    }

    if(cvalue == NULL){
        cvalue = progstr;
    }

    hardware_t hw = NULL;
#ifdef GPIOD
    hardware_t hw = plc_get_hardware(HW_GPIOD); 
    hw->configure(&GpiodConf);
#endif // GPIOD
    Plc = plc_new(
    1, 
    1, 
    0,0,
    N_TIM,
    0,
    N_MEM,
    0,
    STEP,
    hw);

    printf("%s\n", "PLC initialized:");
    dump();
//initialize PLC
    Plc = plc_load_program_file(cvalue, Plc);
//init cli
    Plc = plc_start(Plc);
    for(;;){
        if(Plc->update){
            dump();
            Plc->update=0;
        }
        Plc = plc_func(Plc);
    }    
    return 0;
}

